core/shared/src/test/scala/cats/effect/concurrent/SemaphoreTests.scala

```
@@ -30,7 +30,23 @@ class SemaphoreTests extends AsyncFunSuite with Matchers with EitherValues {
   implicit val cs: ContextShift[IO] = IO.contextShift(executionContext)
   implicit val timer: Timer[IO] = IO.timer(executionContext)
 
+  private def lockSemaphore(n: Long, s: Semaphore[IO]): IO[Boolean] = {
```

I want to understand this method's purpose so that we can try to simplify it (and reduce some of the potential flakiness we have now due to `start` and `timeout`).

AFAICS it concurrently:
- runs an acquire
- loops `count` until there are `n` pending permits.

If that's the case, we can change it so that it concurrently:
- runs an acquire
- loops `count` until there's a negative amount of permits

and then we'll require that amount to be `n`, otherwise fail fast (instead of looping indefinitely with a timeout).

That would look like this:

```scala
private def lockSemaphore(n: Long, s: Semaphore[IO]): IO[Boolean] = {
  s.acquireN(n).background.use(_ => s.count.iterateUntil(_ < 0)).map(n === _)
}
```

No timeouts, so the only non-deterministic thing is how many `s.count` will need to be evaluated. But that's to be expected, as we can't know that upfront...

---

Hey @kubukoz , thanks a lot for the feedback. Nice idea about `background` and `iterateUntil`, I have learned something new 👍 

I have tried that, but because of  `background` calls `cancel`, the result would be released semaphore. 

I reused your idea and the way `background` is implemented. Could you have another look 🙏 ?

---

I think we can return the count from `lockSemaphore` and cancel the acquire, otherwise it's going to run forever :)

I'm not sure why it hangs on JS...

---

Okey, I see the point now. Let me try to re-structure the test.

---

We can't return the count, because the idea of the test to check what happens with Semaphore while it has no available permits. 

I have updated the test to ensure cleaning resources. WDYT, @kubukoz ?